EIPS/eip-3009.md

```diff
@@ -0,0 +1,317 @@
+
+In order to add support for `transferWithAuthorization` to existing contracts ("parent contract") that use the ERC-20 allowance pattern, a forwarding contract ("forwarder") can be constructed that takes an authorization and does the following:
+
+1. Extract the user and deposit amount from the authorization
+2. Call `transferWithAuthorization` to transfer specified funds from the user to the forwarder
+3. Approve the parent contract to spend funds from the forwarder
+4. Call the method on the parent contract that spends the allowance set from the forwarder
+5. Transfer the ownership of any resulting tokens back to the user
+
+**Example:**
+```solidity
+interface IDeFiToken {
+    function deposit(uint256 amount) external returns (uint256);
+
+    function transfer(address account, uint256 amount)
+        external
+        returns (bool);
+}
+
+contract DepositForwarder {
+    bytes4 private constant _TRANSFER_WITH_AUTHORIZATION_SELECTOR = 0xe3ee160e;
+
+    IDeFiToken private _parent;
+    IERC20 private _token;
+
+    constructor(IDeFiToken parent, IERC20 token) public {
+        _parent = parent;
+        _token = token;
+    }
+
+    function deposit(bytes calldata transferAuthorization)
+        external
+        nonReentrant
+        returns (uint256)
+    {
+        (address from, address to, uint256 amount) = abi.decode(
+            transferAuthorization[0:96],
+            (address, address, uint256)
+        );
+        require(to == address(this), "Recipient is not this contract");
+
+        (bool success, ) = address(_token).call(
+            abi.encodePacked(
+                _TRANSFER_WITH_AUTHORIZATION_SELECTOR,
+                transferAuthorization
+            )
+        );
+        require(success, "Failed to transfer to the forwarder");
+
+        require(
+            _token.approve(address(_parent), amount),
+            "Failed to set the allowance"
+        );
+
+        uint256 tokensMinted = _parent.deposit(amount);
+        require(
+            _parent.transfer(from, tokensMinted),
+            "Failed to transfer the minted tokens"
+        );
+
+        uint256 remainder = _token.balanceOf(address(this);
+        if (remainder > 0) {
+            require(
+                _token.transfer(from, remainder),
+                "Failed to refund the remainder"
+            );
+        }
+
+        return tokensMinted;
+    }
+}
+```
+
+## Test Cases
+
+_WIP_
+
+## Implementation
+
+An example implementation can be found in [this repostiory](https://github.com/CoinbaseStablecoin/eip-3009/blob/master/contracts/lib/EIP3009.sol).
+In addition to that, [this file](https://github.com/CoinbaseStablecoin/eip-3009/blob/master/contracts/lib/EIP3009Expanded.sol) shows an expanded form this spec, as implemented by the USDC smart contract.
+
+## Security Considerations
```

Comment:
You should mention/discuss gas-payer re-ordering attacks here.  If I submit two transfers, the gas payer gets to decide what order they occur in.  This could result in an attack opportunity in some cases.

Reply:
Added, thank you.